identity/did/factory.go

```diff
@@ -0,0 +1,149 @@
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	logging "github.com/ipfs/go-log"
+)
+
+var log = logging.Logger("identity")
+
+// Factory is the interface for factory related interactions
+type Factory interface {
+	CreateIdentity(ctx context.Context) (id *DID, err error)
+}
+
+type factory struct {
+	factoryContract *FactoryContract
+	client          ethereum.Client
+	txManager       transactions.Manager
+	queue           *queue.Server
+}
+
+// NewFactory returns a new identity factory service
+func NewFactory(factoryContract *FactoryContract, client ethereum.Client, txManager transactions.Manager, queue *queue.Server) Factory {
+
+	return &factory{factoryContract: factoryContract, client: client, txManager: txManager, queue: queue}
+}
+
+func (s *factory) getNonceAt(ctx context.Context, address common.Address) (uint64, error) {
+	// TODO: add blockNumber of the transaction which created the contract
+	return s.client.GetEthClient().NonceAt(ctx, getFactoryAddress(), nil)
+}
+
+// CalculateCreatedAddress calculates the Ethereum address based on address and nonce
+func CalculateCreatedAddress(address common.Address, nonce uint64) common.Address {
+	// How is a Ethereum address calculated:
+	// See https://ethereum.stackexchange.com/questions/760/how-is-the-address-of-an-ethereum-contract-computed
+	return crypto.CreateAddress(address, nonce)
+}
+
+func (s *factory) createIdentityTX(opts *bind.TransactOpts) func(accountID id.CentID, txID uuid.UUID, txMan transactions.Manager, errOut chan<- error) {
+	return func(accountID id.CentID, txID uuid.UUID, txMan transactions.Manager, errOut chan<- error) {
+		ethTX, err := s.client.SubmitTransactionWithRetries(s.factoryContract.CreateIdentity, opts)
+		if err != nil {
+			errOut <- err
+			log.Infof("Failed to send identity for creation [txHash: %s] : %v", ethTX.Hash(), err)
+			return
+		}
+
+		log.Infof("Sent off identity creation Ethereum transaction hash [%x] and Nonce [%v] and Check [%v]", ethTX.Hash(), ethTX.Nonce(), ethTX.CheckNonce())
+		log.Infof("Transfer pending: 0x%x\n", ethTX.Hash())
+
+		res, err := ethereum.QueueEthTXStatusTask(accountID, txID, ethTX.Hash(), s.queue)
+		if err != nil {
+			errOut <- err
+			return
+		}
+
+		_, err = res.Get(txMan.GetDefaultTaskTimeout())
+		if err != nil {
+			errOut <- err
+			return
+		}
+		errOut <- nil
+	}
+
+}
+
+func (s *factory) calculateIdentityAddress(ctx context.Context) (*common.Address, error) {
+	factoryAddress := getFactoryAddress()
+	nonce, err := s.getNonceAt(ctx, factoryAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	identityAddress := CalculateCreatedAddress(factoryAddress, nonce)
+	log.Infof("Calculated Address of the identity contract: 0x%x\n", identityAddress)
+	return &identityAddress, nil
+}
+
+func (s *factory) isIdentityContract(identityAddress common.Address) error {
+	contractCode, err := s.client.GetEthClient().CodeAt(context.Background(), identityAddress, nil)
```

Comment:
Do we know in the case of upgradable smart contracts, if this returns the proxy contract or the abi implementation one? Otherwise this will not work once we support zos upgradable smart contracts

Reply:
As discussed. Identity doesn't have a proxy contract.

Anyway would be interesting.

@rdinicut do you know?

What would `eth.getCode` return in case of upgradeability?

Reply:
it will return the bytecode without the contructor and initialisation code

Reply:
I did not test that but I am assuming it will return the bytecode from the Proxy Zos contract,